### PR TITLE
Estimated remediation date

### PR DESCRIPTION
**Pull request by 37b**
This PR adds a new field to the Finding model to store the Planned Remediation Date. This is used to track progress towards remediation and to indicate to security teams that the Finding is being triaged. This is useful to organizations without Jira.

### CODE REVIEW

**Comment by 37b** [comment]
Closing while I read more on migrations

**Comment by 37b** [comment]
I am still trying to figure out the best way to get the migration file from the container to version control. I don't expect this PR will be completed, but am looking for feedback on the change and the approach.

**Comment by Maffooch** [comment]
@37bIf you are using the `dev` environment, then the working directory is bind mounted. Once you make the migrations from the uwsgi container, it will also be created on the host machine and be in version control. 

For this PR though, there is a few extra fields in the migration that I am not sure how they made it into. Beyond that,  and some of the formatting changes, the PR looks good overall!

**Comment by 37b** [comment]
> @37b If you are using the `dev` environment, then the working directory is bind mounted. Once you make the migrations from the uwsgi container, it will also be created on the host machine and be in version control.
> 
> For this PR though, there is a few extra fields in the migration that I am not sure how they made it into. Beyond that, and some of the formatting changes, the PR looks good overall!

@cody-m-tibco 

Thanks for the info. I was having trouble running locally and knew it was related to the M1 Mac (ARM) but didn't have time to look into a fix. It was pretty simple - just needed to add `profile: linux/amd64` to the compose file. I'll redo the migrations and fix failing tests.

**Comment by 37b** [comment]
@cody-m-tibco  Are some of these tests broken overall? I can't get them to pass using a clean `dev` branch

**Comment by Maffooch** [comment]
@37b I looked at each of the failing tests and it is the same test at each level that is failing. It is when creating a finding manually. I image selenium is getting hung up on the extra field on the page. You can run the tests manually by following [this guide](https://github.com/DefectDojo/django-DefectDojo/blob/master/readme-docs/DOCKER.md#running-the-integration-tests)

I also recommend commenting out [this line](https://github.com/DefectDojo/django-DefectDojo/blob/d55fb295e3eae8167af320c545445b2098422737/tests/base_test_class.py#L64) to see where the hang up is directly on the page

**Comment by 37b** [comment]
@Maffooch  Thanks for your help with the tests - I am still getting my bearings with Django development.

**Comment by devGregA** [comment]
@37b targeting next release for getting this merged. Thank you for your PR!

File: dojo/forms.py [hunk 1]
```diff
@@ -3195,7 +3218,62 @@ def __init__(self, *args, **kwargs):
         self.group = kwargs.pop('group', None)
         super(ConfigurationPermissionsForm, self).__init__(*args, **kwargs)
 
-        self.permission_fields = get_configuration_permissions_fields()
+        if get_system_setting('enable_github'):
```
Comment (StefanFl) [inline dojo/forms.py:3221]: You have to rebase the branch to the actual dev branch. Here you re-introduce an old version of some code.

File: adrs/001-planned-remediation-date.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# 001 - Add new Planned Remediation Date field to Finding model
```
Comment (StefanFl) [inline adrs/001-planned-remediation-date.md:1]: What is the thought of other maintainers? Having such a document does only make sense, if we want to continue writing adrs. Otherwise its a singular thing. I personally wouldn't need it and I can't imagine forcing other contributers to write these documents.

File: docker-compose.override.unit_tests.yml [hunk 3]
```diff
@@ -39,6 +39,7 @@ services:
     image: busybox:1.35.0-musl
     entrypoint: ['echo', 'skipping', 'initializer']
   mysql:
+    platform: linux/amd64
```
Comment (StefanFl) [inline docker-compose.override.unit_tests.yml:42]: How is this related to the PR?

File: dojo/forms.py [hunk 4]
```diff
@@ -3231,3 +3309,63 @@ def set_permission(self, codename):
                 self.group.auth_group.permissions.remove(self.permissions[codename])
             else:
                 raise Exception('Neither user or group are set')
+
+
+class Permission_Helper:
```
Comment (StefanFl) [inline dojo/forms.py:3314]: This again is code that is not related to the PR.

File: dojo/models.py [hunk 5]
```diff
@@ -2062,6 +2062,11 @@ class Finding(models.Model):
                                verbose_name=_('Service'),
                                help_text=_('A service is a self-contained piece of functionality within a Product. This is an optional field which is used in deduplication of findings when set.'))
 
+    planned_remediation_date = models.DateField(null=True,
+                                                editable=True,
+                                                verbose_name=_('Planned Remediation Date'),
+                                                help_text=_("The date the flaw is expected to be remediated."))
```
Comment (StefanFl) [inline dojo/models.py:2068]:
```suggestion
                                                help_text=_("The date the finding is expected to be remediated."))
```

We never use the word flaw, so it should be finding to be consistent

File: tests/base_test_class.py [hunk 6]
```diff
@@ -1,12 +1,14 @@
+import os
```
Comment (StefanFl) [inline tests/base_test_class.py:1]: There is no change in this file apart from the order of imports. Makes reviewing the PR unnecessarily more work.

File: tests/product_test.py [hunk 7]
```diff
@@ -1,9 +1,11 @@
-from selenium.webdriver.common.by import By
-from selenium.webdriver.support.ui import Select
-from selenium.webdriver.common.keys import Keys
-import unittest
 import sys
 import time
+import unittest
```
Comment (StefanFl) [inline tests/product_test.py:3]: See before, another file with no meaningful changes.

**Review by StefanFl — CHANGES_REQUESTED** [review]
Hi @37b, sorry it took a while until someone reviewed your PR. I found some needed changes in the code but will checkout and try it as well now.

File: dojo/templates/dojo/findings_list_snippet.html [hunk 8]
```diff
@@ -219,6 +222,7 @@ <h3 class="has-filters">
                             <th class="nowrap">{% dojo_sort request 'Product' 'test__engagement__product__name'%}</th>
                             {% endif %}
                             <th>Service</th>
+                            <th>Planned Remediation Date</th>
```
Comment (StefanFl) [inline dojo/templates/dojo/findings_list_snippet.html:225]: The list has already a lot of fields and this title needs much more space than the actual date. Can we just write "Planned Remediation"?

File: dojo/templates/dojo/view_finding.html [hunk 9]
```diff
@@ -236,6 +244,9 @@ <h3 class="pull-left finding-title">
                     {% if finding.publish_date %}
                         <th>Vuln Publish date</th>
                     {% endif %}
+                    {% if finding.planned_remediation_date %}
+                        <th>Planned Remediation Date</th>
```
Comment (StefanFl) [inline dojo/templates/dojo/view_finding.html:248]: This area has already a lot of fields and this title needs much more space than the actual date. Can we just write "Planned Remediation"?

File: dojo/templates/dojo/view_test.html [hunk 10]
```diff
@@ -678,6 +681,7 @@ <h4 class="has-filters"> Findings ({{findings.total_count}}) <small>{{ test.id|g
                             {% if 'is_finding_groups_enabled'|system_setting_enabled %}
                                 <th>Group</th>
                             {% endif %}
+                            <th>Planned Remediation Date</th>
```
Comment (StefanFl) [inline dojo/templates/dojo/view_test.html:684]: The list has already a lot of fields and this title needs much more space than the actual date. Can we just write "Planned Remediation"?

File: dojo/templates/dojo/findings_list_snippet.html [hunk 11]
```diff
@@ -91,6 +91,9 @@ <h3 class="has-filters">
                                 <option value="Critical">Critical</option>
                             </select><br/><br/>
 
+                            <label><b>Planned Remediation Date</b><br/>
+                                <input id="id_bulk_planned_remediation_date" name="planned_remediation_date" type="date" alt="Select the date the Finding will be remediated"/></label><br/>
```
Comment (StefanFl) [inline dojo/templates/dojo/findings_list_snippet.html:95]: The placeholder **TT.MM.JJJJ** is bold and black and doesn't look very nice. Can we have this in a normal font with a grey color?

File: dojo/templates/dojo/view_test.html [hunk 12]
```diff
@@ -551,6 +551,9 @@ <h4 class="has-filters"> Findings ({{findings.total_count}}) <small>{{ test.id|g
                             <option value="High">High</option>
                             <option value="Critical">Critical</option>
                         </select><br/><br/>
+                        <label><b>Planned Remediation Date</b><br/>
+                        <input id="id_bulk_planned_remediation_date" name="planned_remediation_date" type="date"
```
Comment (StefanFl) [inline dojo/templates/dojo/view_test.html:555]: The placeholder **TT.MM.JJJJ** is bold and black and doesn't look very nice. Can we have this in a normal font with a grey color?

**Review by StefanFl — CHANGES_REQUESTED** [review]

File: dojo/templates/dojo/findings_list_snippet.html (hunk 8 above)
Comment (37b) [reply dojo/templates/dojo/findings_list_snippet.html]: Fine with me!

File: dojo/templates/dojo/view_finding.html (hunk 9 above)
Comment (37b) [reply dojo/templates/dojo/view_finding.html]: Fine with me!

File: tests/base_test_class.py (hunk 6 above)
Comment (37b) [reply tests/base_test_class.py]: That was an IDE thing - I thought I reverted those. Will fix.

File: tests/product_test.py (hunk 7 above)
Comment (37b) [reply tests/product_test.py]: That was an IDE thing - I thought I reverted those. Will fix.

File: adrs/001-planned-remediation-date.md (hunk 2 above)
Comment (37b) [reply adrs/001-planned-remediation-date.md]: OK Just thought it would be useful. I'll remove it as well.

File: docker-compose.override.unit_tests.yml (hunk 3 above)
Comment (37b) [reply docker-compose.override.unit_tests.yml]: Again, this was introduced by accident while developing on an ARM64 Mac - apparently there is no arm64 mysql so I had to specifically request the amd64. Will remove.

File: dojo/templates/dojo/findings_list_snippet.html (hunk 11 above)
Comment (37b) [reply dojo/templates/dojo/findings_list_snippet.html]: Looks like a default Django forms thing. I'll dig into it.

File: dojo/templates/dojo/view_test.html (hunk 12 above)
Comment (37b) [reply dojo/templates/dojo/view_test.html]: Looks like a default Django forms thing. I'll dig into it.

**Comment by 37b** [comment]
@StefanFl @Maffooch Finally got this going, can you give it another look?

**Comment by Maffooch** [comment]
Looks okay to me after the compose file is removed from the commit

@StefanFl what do you think?

**Review by Maffooch — APPROVED** [review]

**Review by blakeaowens — APPROVED** [review]
Looks good to me.
